java-21/Dockerfile

```diff
@@ -46,33 +46,48 @@ jdk.zipfs\
 # Extract java SSL certs
 FROM gcr.io/distroless/java-base-debian12:debug AS ssl
 
-FROM amd64/debian:bookworm-slim As builder
+FROM debian:bookworm-slim AS builder
 RUN apt update && apt install -y libjemalloc-dev
 
-# Debian-12 image
-FROM gcr.io/distroless/cc-debian12:debug
-
-MAINTAINER Hypertrace "https://www.hypertrace.org/"
+# stage for amd64
+FROM gcr.io/distroless/cc-debian12:debug AS base-amd64
 
 SHELL ["/busybox/sh", "-c"]
-
 RUN ln -s /busybox/sh /bin/sh
 
-COPY --from=ssl /etc/ssl/certs/java /etc/ssl/certs/java
-
 COPY --from=jre /lib/x86_64-linux-gnu/libz.so.1.2.11 /lib/x86_64-linux-gnu/libz.so.1.2.11
 RUN ln -s /lib/x86_64-linux-gnu/libz.so.1.2.11 /lib/x86_64-linux-gnu/libz.so.1
 
-COPY --from=jre /jre /usr/lib/jvm/zulu-21-amd64-slim
-RUN ln -s /usr/lib/jvm/zulu-21-amd64-slim/bin/java /usr/bin/java
-RUN ln -s /usr/lib/jvm/zulu-21-amd64-slim/bin/jfr /usr/bin/jfr
-
-# change rocksdb default memory allocator
-# https://docs.confluent.io/platform/current/streams/developer-guide/memory-mgmt.html#rocksdb
+# change default memory allocator to jemalloc
 COPY --from=builder /usr/lib/x86_64-linux-gnu/libjemalloc* /usr/lib/x86_64-linux-gnu/
 ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so
 
+# stage for arm64
+FROM gcr.io/distroless/cc-debian12:debug AS base-arm64
+
+SHELL ["/busybox/sh", "-c"]
+RUN ln -s /busybox/sh /bin/sh
+
+COPY --from=jre /lib/aarch64-linux-gnu/libz.so.1.2.11 /lib/aarch64-linux-gnu/libz.so.1.2.11
+RUN ln -s /lib/aarch64-linux-gnu/libz.so.1.2.11 /lib/aarch64-linux-gnu/libz.so.1
+
+# change default memory allocator to jemalloc
+COPY --from=builder /usr/lib/aarch64-linux-gnu/libjemalloc* /usr/lib/aarch64-linux-gnu/
+ENV LD_PRELOAD=/usr/lib/aarch64-linux-gnu/libjemalloc.so
+
+# final image
+ARG TARGETARCH
+FROM base-${TARGETARCH}
+
+MAINTAINER Hypertrace "https://www.hypertrace.org/"
+
+COPY --from=ssl /etc/ssl/certs/java /etc/ssl/certs/java
+
+COPY --from=jre /jre /usr/lib/jvm/zulu-21-slim
+RUN ln -s /usr/lib/jvm/zulu-21-slim/bin/java /usr/bin/java
+RUN ln -s /usr/lib/jvm/zulu-21-slim/bin/jfr /usr/bin/jfr
+
 # set JAVA_HOME
-ENV JAVA_HOME=/usr/lib/jvm/zulu-21-amd64-slim
+ENV JAVA_HOME=/usr/lib/jvm/zulu-21-slim
```

Comment:
This change alone isn't really changing anything is it? Since no one is building the base docker images, and the CI job here hasn't been changed we're still going to end up only publishing the amd64 version right?

If I'm following that correctly, can we update to publish this as multiarch too?